lib/base/io-engine.hpp

```diff
@@ -163,51 +166,73 @@ class AsioConditionVariable
 /**
  * I/O timeout emulator
  *
+ * This class provides a workaround for Boost.ASIO's lack of built-in timeout support.
+ * While Boost.ASIO handles asynchronous operations, it does not natively support timeouts for these operations.
+ * This class uses a boost::asio::deadline_timer to emulate a timeout by scheduling a callback to be triggered
+ * after a specified duration, effectively adding timeout behavior where none exists.
+ * The callback is executed within the provided strand, ensuring thread-safety.
+ *
+ * The constructor returns immediately after scheduling the timeout callback. The callback itself is invoked
+ * asynchronously when the timeout occurs, but control is immediately returned to the caller without waiting for the
+ * timeout to expire. This allows the caller to continue execution while the timeout is running in the background.
+ *
+ * The class provides a Cancel() method to unschedule any pending callback. If the callback has already been run,
+ * calling Cancel() has no effect. This method can be used to abort the timeout early if the monitored operation
+ * completes before the callback has been run. The Timeout destructor also automatically cancels any pending callback.
+ * A callback is considered pending even if the timeout has already expired,
+ * but the callback has not been executed yet due to a busy strand.
+ *
  * @ingroup base
  */
 class Timeout : public SharedObject
```

Comment:
Do we still need `Timer` to inherit from `SharedObject`? Shouldn't allocating `Timer` on stack be just fine (just like you would do with ` std::unique_lock`)?

Reply:
Shall I also remove unnecessary explicit Cancel() while on it?

Reply:
Done it locally. Now no Cancel() usages are left ex. destructor and unit tests. 😁